app/src/main/java/ch/epfl/qedit/view/treasurehunt/QuestionLocatorActivity.java

```diff
@@ -0,0 +1,189 @@
+
+        // We retrieve the location from the activity arguments
+        Bundle bundle = Objects.requireNonNull(getIntent().getExtras());
+        questionLoc = (Location) bundle.getParcelable(QUESTION_LOCATION);
+        questionRadius = bundle.getFloat(QUESTION_RADIUS);
+
+        // We retrieve the UI elements and set up the default UI
+        text1 = findViewById(R.id.question_locator_text1);
+        text2 = findViewById(R.id.question_locator_text2);
+        button = findViewById(R.id.question_locator_button);
+        setUnknownUI();
+
+        // We retrieve the two singleton backend services
+        locService = LocServiceFactory.getInstance(getApplicationContext());
+        permManager = PermManagerFactory.getInstance();
+
+        // If we couldn't subscribe, we need to ask for permissions for the location
+        if (!locService.subscribe(this, LOCATION_INTERVAL)) askPermissions();
+    }
+
+    private void askPermissions() {
+        permManager.requestPermissions(this, this, REQUESTED_PERMISSIONS);
+    }
+
+    @Override
+    public void onPermissionResult(String[] permissions, boolean[] granted) {
+        for (int i = 0; i < granted.length; i++) {
+            if (!granted[i]) {
+                // If the permissions were not granted, we just show an error UI to the user
+                // If it is the first time the user denies, we show the button, otherwise we
+                // just give up and show nothing.
+                if (permManager.shouldAskAgain(this, permissions[i])) setPermissionUI();
+                else button.setVisibility(View.GONE);
+
+                return;
+            }
+        }
+
+        // The user finally accepted, so we can hide the button and subscribe to the location
+        locService.subscribe(this, LOCATION_INTERVAL);
+        button.setVisibility(View.GONE);
+    }
+
+    // Sets the text of the UI to inform the user that his position is unknown
+    private void setUnknownUI() {
+        text1.setText(getString(R.string.question_locator_error));
+        text2.setText("");
+    }
+
+    // Sets the button of the UI so that the user can allow the location
+    private void setPermissionUI() {
+        button.setOnClickListener(v -> askPermissions());
+        button.setText(R.string.enable_location);
+        button.setVisibility(View.VISIBLE);
+    }
+
+    // This method shows the final UI that appears when the user found the question
+    private void setFinishUI() {
+        text1.setText(getString(R.string.question_locator_found));
+        text2.setText("");
+        button.setOnClickListener(
+                v -> Toast.makeText(this, "Not yet implemented", Toast.LENGTH_SHORT).show());
+        button.setText(getString(R.string.question_locator_answer));
+        button.setVisibility(View.VISIBLE);
+    }
+
+    // This method is used to prevent the modulo from returning negative numbers
+    private static float mod(float a, float b) {
+        float m = a % b;
```

Comment:
Why don't you use `Math.floorMod()` ?

Reply:
Very good, point, I forgot about that function !

Reply:
Unfortunately, `Math.floorMod` only works for integer types, so I couldn't use it. I could cast to an  int since the user doesn't care about the decimals, but I am not so keen on doing that...